stumpy/stump.py

```diff
@@ -177,38 +198,49 @@ def _compute_diagonal(
                 #     - (T_B[i + k - 1] - M_T_m_1[i + k]) * (T_A[i - 1] - μ_Q_m_1[i])
                 # )
                 cov = cov + constant * (
-                    cov_a[i + k] * cov_b[i] - cov_c[i + k] * cov_d[i]
+                    cov_a[i + g] * cov_b[i] - cov_c[i + g] * cov_d[i]
                 )
 
-            if T_B_subseq_isfinite[i + k] and T_A_subseq_isfinite[i]:
+            if T_B_subseq_isfinite[i + g] and T_A_subseq_isfinite[i]:
                 # Neither subsequence contains NaNs
-                if T_B_subseq_isconstant[i + k] or T_A_subseq_isconstant[i]:
+                if T_B_subseq_isconstant[i + g] or T_A_subseq_isconstant[i]:
                     pearson = 0.5
                 else:
-                    pearson = cov * Σ_T_inverse[i + k] * σ_Q_inverse[i]
+                    pearson = cov * Σ_T_inverse[i + g] * σ_Q_inverse[i]
 
-                if T_B_subseq_isconstant[i + k] and T_A_subseq_isconstant[i]:
+                if T_B_subseq_isconstant[i + g] and T_A_subseq_isconstant[i]:
                     pearson = 1.0
 
                 if pearson > ρ[thread_idx, i, 0]:
```

Comment:
That would be a good idea! I will do so and update you here.

Reply:
Thank you

Reply:
> an you please compare the performance of `k=1` with the original version with `T = np.ranom.rand(100000)` and `m=50`? I hope that this new version adds only negligible computational time. Then, compare the performance for different `k in range(1, 11)`?


Computing times (in sec) are as follows:

```
k = 1
--> original stump: 33   
--> top-1 stump: 35

###

k = 2    --> 21
k = 3    --> 26
k = 4    --> 33
k = 5    --> 38
k = 6    --> 44
k = 7    --> 49
k = 8    --> 55
k = 9    --> 63
k = 10   --> 69
```

I do not know why the computing time for `k = 1` is higher than the cases `k=2` and `k=3`. To be on the safe side, I did the test twice to make sure the numbers are correct.

Reply:
Can you please share your testing code with me?

Reply:
like...here? Please see below:

```
import numpy as np
import stumpy
import time
###
seed = 0
np.random.seed(seed)
T = np.random.rand(100000)
m = 50
###
lst_ComputingTime =  []
for k in range(1, 11):
    tic = time.time()
    out = stumpy.stump(T, m, k=k)
    toc = time.time()
    
    lst_ComputingTime.append(toc - tic)
```

Reply:
Ahh, okay. So, one thing you need to remember/know is that the first time a `numba` function is called (within a script) it needs to be JIT-compiled. Then, subsequent calls to that same function do not need to be compiled. Thus, whenever you are doing timing comparisons, you should always call the function once and throw away the timing as the timing for the first run will include compilation time (likely 2-3 seconds, sometimes longer):

```
import numpy as np
import stumpy
import time
###
seed = 0
np.random.seed(seed)
T = np.random.rand(100000)
m = 50
###

stumpy.stump(T, m, k=1)  # throw away, compile once

lst_ComputingTime =  []
for k in range(1, 11):
    tic = time.time()
    out = stumpy.stump(T, m, k=k)
    toc = time.time()
    
    lst_ComputingTime.append(toc - tic)
```

> And, for the original version: out = stumpy.stump(T, m)

Does this mean that you re-installed STUMPY without all of your changes and executed the timings script?

Reply:
@seanlaw  
Do you want me to conduct the performance test again?

Reply:
Yes, please!

Reply:
```
k = 1
--> original stump: 18  (note: I did ` # throw away, compile once` here as well)
--> top-1 stump: 15

###
k = 1    --> 15 (top-1 stump)
k = 2    --> 19
k = 3    --> 25
k = 4    --> 31
k = 5    --> 39
k = 6    --> 51
k = 7    --> 64
k = 8    --> 57
k = 9    --> 63
k = 10   --> 69
```

Reply:
Hmm, I am a little suspicious as to why original stump is slower by 3 seconds compared with top-1. I would've expected them to be virtually identical given that they have the same input time series.

Additionally, I think it is wise to add an additional note to the docstrings for `k` that it would require additional computational time and space (memory).

Reply:
@seanlaw 
Do you know how we can investigate why this happens? (Or, should we ignore it?)

Maybe considering different variables for left and right profile affects the performance of Numba(?) so, I can change the code so that we have one variable for the matrix profile (mp) and the mp indices, and see how it performs. 

What do you think?

Reply:
I don't know. I mean, I can't complain if `top-1` is faster than the current implementation. However, my concern is that there is something "wrong". Having a faster time implies that we are no longer computing "something" in `top-1` that we did in the original. I'm just trying not to be naive. I think it is worth understanding where the sudden speed gain came from. Initially, I thought that it might have been due to the input array but if both inputs are the same (with the same seed) then I really don't expect things to be drastically different.

Reply:
> However, my concern is that there is something "wrong". Having a faster time implies that we are no longer computing "something" in top-1 that we did in the original. I'm just trying not to be naive. I think it is worth understanding where the sudden speed gain came from | Figuring this out would require a bit of trial-and-error

I understand. Please allow me some time to see if I can figure it out.

Reply:
@seanlaw 
So, I think I found the source of speed gain. 

I created a new version of original `stump` in which I just considered separate variables for left and right Pearson correlation / matrix profile/ matrix profile indices (similar to what we did in top-k). In below, I am providing the performance of different versions of `stump`:

```
original stump :
17.9

original stump, but with separate variables for PL, PR, IL, IR:
14.3

Top-k stump (with k=1):
14.8
```

I still do not know why this separation of variables speed up the computation. Thoughts?

Reply:
Oh, wow! That's pretty wild and thank you for looking into it. I would've never expected that pulling those things out would make things faster. Consider this to be resolved

Reply:
>  I would've never expected that pulling those things out would make things faster.

Yes. That surprised me! 

> Consider this to be resolved

Great. Thanks for being cautious about this. That was a valuable lesson!